### PR TITLE
OOP Student Library - Add Student, Teacher and Person classes

### PR DESCRIPTION
**Pull request by aamirkhan2478**
### Project requirements
- Class `Person` :
  - Instance vars: `@id`, `@name`, and `@age`.
  - Constructor with `name`,  `age`, and `parent_permission` as parameters. `name` and `parent_permission` are optional and have default values of `"Unknown"` and `true`.
  - Getters for `@id`, `@name`, and `@age`.
  - Setters for `@name` and `@age`.
  - Private method `is_of_age?` that returns `true` if `@age` is greater or equal to 18 and `false` otherwise.
  - Public method `can_use_services?` that returns `true` if a person is of age or if they have permission from parents.
- Class `Student`:
  - Inherits from `Person`.
  - Constructor extends parent's constructor by adding `@classroom` and a parameter for it.
  - Method `play_hooky` that returns `"¯\(ツ)/¯"`.
- Class `Teacher`:
  - Inherits from `Person`.
  - Constructor extends parent's constructor by adding `@specialization` and a parameter for it.
  - Override `can_use_services?` so it always returns `true`.

### CODE REVIEW

File: person.rb [hunk 1]
```diff
@@ -0,0 +1,20 @@
+class Person
+  attr_accessor :name, :age
+  attr_reader :id
+
+  def initialize(name, age = 'Unknown', parent_permission = 'true')
+    @name = name
+    @age = age
+    @parent_permission = parent_permission
+  end
```
Comment (mikethreels) [inline person.rb]: - [ ] in the Person class the instance variable `@id` should be created as well.

File: person.rb [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+  def can_use_services?
+    is_of_age? || parent_permission
```
Comment (mikethreels) [inline person.rb]: - [ ] you are calling the method `is_of_age?` here however the actual method is called `of_age?`

File: student.rb [hunk 3]
```diff
@@ -0,0 +1,11 @@
+require './person'
+class Student < Person
+  def initialize(classroom)
+    super()
```
Comment (mikethreels) [inline student.rb]: - [ ] Constructor should extend the parent's constructor by adding @classroom and a parameter for it. please have a look at [this link ](https://github.com/microverseinc/curriculum-ruby/blob/main/oop/articles/class_and_object_for_irb.md#constructor-and-instance-vars)for an example

File: teacher.rb [hunk 4]
```diff
@@ -0,0 +1,12 @@
+require './person'
+
+class Teacher < Person
+  def initialize(specialization)
+    super()
+    @specialization = specialization
+  end
```
Comment (mikethreels) [inline teacher.rb]: - [ ] Constructor should extend the parent's constructor by adding @specialization and a parameter for it. please have a look at [this link ](https://github.com/microverseinc/curriculum-ruby/blob/main/oop/articles/class_and_object_for_irb.md#constructor-and-instance-vars)for an example

**Review by mikethreels — CHANGES_REQUESTED** [review]
Hi Aamir, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## Highlights
- linters are passing

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: student.rb [hunk 5]
```diff
@@ -1,11 +1,12 @@
-require './person'
+require "./person"
+
 class Student < Person
-  def initialize(classroom)
-    super()
+  def initialize(age, name, parent_permission, classroom)
+    super(age, name, parent_permission)
```
Comment (ShadyShawkat) [inline student.rb]: - [ ] Kindly fix the order of the passed parameters to the `super()` method. Currently, this is how your `Person` class's constructor looks like:
```ruby
def initialize(name, age = "Unknown", parent_permission = "true")
```
As you can see, you pass in the parameters in a different order than when you passed it in the `super()` method.

File: teacher.rb [hunk 6]
```diff
@@ -1,8 +1,8 @@
-require './person'
+require "./person"
 
 class Teacher < Person
-  def initialize(specialization)
-    super()
+  def initialize(age, name, specialization, parent_permission='true')
+    super(age, name, parent_permission)
```
Comment (ShadyShawkat) [inline teacher.rb]: - [ ] Kindly fix the order of the passed parameters to the `super()` method. Currently, this is how your `Person` class's constructor looks like:
```ruby
def initialize(name, age = "Unknown", parent_permission = "true")
```
As you can see, you pass in the parameters in a different order than when you passed it in the `super()` method.

**Review by ShadyShawkat — CHANGES_REQUESTED** [review]
## Hi @aamirkhan2478, 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't been addressed yet.

Image

 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by gemmen29 — APPROVED** [review]
Hi @aamirkhan2478, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## Highlights 💯 
- No linters errors ✔️ 
- Descriptive pull request ✔️ 
- Descriptive commit messages ✔️ 



## Optional suggestions

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**